File: C++/792. Number of Matching Subsequences.cpp
class Solution {
public:
    bool is(string s,string t){
        int i,f=0;
        for(i=0;i<s.size()&&f < t.size();i++){
            if(s[i]==t[f])
            f++;
        }
        return f==t.size();
    }
    int numMatchingSubseq(string s, vector<string>& w) {
        int i,c=0;
        map<string,bool>a;
        for(i=0;i<w.size();i++){
            if(a.find(w[i])!=a.end()){
                if(a[w[i]])
                c++;
            }
            else{
                bool r=is(s,w[i]);
                a[w[i]]=r;
                if(r)
                c++;
            }
        }
        return c;
    }
};

File: JAVA/792. Number of Matching Subsequences.java
class Solution {
    public int numMatchingSubseq(String s, String[] w) {
        int i,c=0;
        HashMap<String, Boolean> a = new HashMap<>();
        for(i=0;i<w.length;i++){
            if(a.containsKey(w[i])) {
                if(a.get(w[i])) c++;
            } else {
                boolean result = is(s, w[i]);
                a.put(w[i], result);
                if(result) c++;
            }
        }
        return c;
    }
    private boolean is(String s,String t){
        int i,f=0;
        for(i=0;i<s.length() && f<t.length();i++){
            if(s.charAt(i)==t.charAt(f))
            f++;
        }
        return f==t.length();
    }
}